crates/algos/src/dss.rs

```diff
@@ -0,0 +1,203 @@
+                .collect::<Vec<_>>()
+                .into_boxed_slice(),
+        )
+    }
+
+    /// Returns the number of elements in the dss, also referred to
+    /// as its 'length'.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use graph::dss::DisjointSetStruct;
+    ///
+    /// let dss = DisjointSetStruct::<usize>::new(3);
+    /// assert_eq!(dss.len(), 3);
+    /// ```
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Joins the set of `id1` with the set of `id2`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use graph::dss::DisjointSetStruct;
+    ///
+    /// let dss = DisjointSetStruct::new(10);
+    /// dss.union(2, 4);
+    /// assert_eq!(dss.find(2), 2);
+    /// assert_eq!(dss.find(4), 2);
+    /// ```
+    pub fn union(&self, mut id1: NI, mut id2: NI) {
+        loop {
+            id1 = self.find(id1);
+            id2 = self.find(id2);
+
+            if id1 == id2 {
+                return;
+            }
+
+            // We do Union-by-Min, so the smaller set id wins.
+            // We also only update the entry for id1 and if that
+            // is the smaller value, we need to swap ids so we update
+            // only the value for id2, not id1.
+            if id1 < id2 {
+                let tmp = id2;
+                id2 = id1;
+                id1 = tmp;
+            }
+
+            let old_entry = id1;
+            let new_entry = id2;
+
+            if self.update_parent(id1, old_entry, new_entry).is_ok() {
+                break;
+            }
+        }
+    }
+
+    /// Find the set of `id`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use graph::dss::DisjointSetStruct;
+    ///
+    /// let dss = DisjointSetStruct::new(10);
+    /// assert_eq!(dss.find(4), 4);
+    /// dss.union(4, 2);
+    /// assert_eq!(dss.find(4), 2);
+    /// ```
+    pub fn find(&self, mut id: NI) -> NI {
+        let mut parent = self.parent(id);
+
+        while id != parent {
+            let grand_parent = self.parent(parent);
+            // Try to apply path-halving by setting the value
+            // for some id to its grand parent. This might fail
+            // if another thread is also changing the same value
+            // but that's ok. The CAS operations guarantees
+            // that at least one of the contenting threads will
+            // succeed. That's enough for the path-halving to work
+            // and there is no need to retry in case of a CAS failure.
+            self.update_parent(id, parent, grand_parent).unwrap();
+            id = parent;
+            parent = grand_parent;
+        }
+
+        id
+    }
+
+    #[inline]
+    fn parent(&self, i: NI) -> NI {
+        self.0[i.index()].load(Ordering::Acquire)
+    }
+
+    #[inline]
+    fn update_parent(&self, id: NI, current: NI, new: NI) -> Result<NI, NI> {
+        self.0[id.index()].compare_exchange_weak(current, new, Ordering::Release, Ordering::Relaxed)
```

Comment:
probably want to use at least Acquire ordering for the success load, otherwise we might never get to do any path halving. (Note that all three references use the equivalent of SeqCst here). I think Relaxed for the failure load is fine, as we're not doing anything with that value.

```suggestion
        self.0[id.index()].compare_exchange_weak(current, new, Ordering::AcqRel, Ordering::Relaxed)
```